File: java/decouple-monolith/exercise/src/main/java/payments/temporal/PaymentsWorkerApp.java
package payments.temporal;

import compliance.ComplianceChecker;
import compliance.temporal.activity.ComplianceActivityImpl;
import io.temporal.client.WorkflowClient;
import io.temporal.client.WorkflowClientOptions;
import io.temporal.serviceclient.WorkflowServiceStubs;
import io.temporal.worker.Worker;
import io.temporal.worker.WorkerFactory;
import io.temporal.worker.WorkflowImplementationOptions;
import io.temporal.workflow.NexusServiceOptions;
import payments.PaymentGateway;
import payments.Shared;
import payments.temporal.activity.PaymentActivityImpl;

import java.util.Collections;

/**
 * ═══════════════════════════════════════════════════════════════════
 *  MONOLITH VERSION — This works! Run it at Checkpoint 0.
 * ═══════════════════════════════════════════════════════════════════
 *
 * Currently this single worker handles EVERYTHING:
 *   - PaymentProcessingWorkflow
 *   - PaymentActivity (validate + execute)
 *   - ComplianceActivity (compliance check)  <-- will move to its own worker
 *
 * All on one task queue: "payments-processing"
 *
 * ── TODO 5: Add Nexus endpoint mapping + remove ComplianceActivity ──
 *
 * After completing TODOs 1-4, come back here and make THREE changes:
 *
 * CHANGE 1: Register both workflows with NexusServiceOptions (maps service to endpoint)
 *   Currently:   worker.registerWorkflowImplementationTypes(PaymentProcessingWorkflowImpl.class)
 *   Change to:   worker.registerWorkflowImplementationTypes(
 *                    WorkflowImplementationOptions.newBuilder()
 *                        .setNexusServiceOptions(Collections.singletonMap(
 *                            "ComplianceNexusService",
 *                            NexusServiceOptions.newBuilder()
 *                                .setEndpoint("compliance-endpoint")
 *                                .build()))
 *                        .build(),
 *                    PaymentProcessingWorkflowImpl.class,
 *                    ReviewCallerWorkflowImpl.class);
 *
 * CHANGE 2: Remove ComplianceActivityImpl registration
 *   The compliance check now runs on the Compliance worker via Nexus.
 *   Delete the lines that register ComplianceActivityImpl.
 *   Also remove the ComplianceChecker import and instantiation.
 *
 * CHANGE 3: (Already done) ReviewCallerWorkflowImpl is registered above.
 *   Both workflows use the same ComplianceNexusService endpoint mapping.
 */
public class PaymentsWorkerApp {

    public static void main(String[] args) {
        // C — Connect to Temporal (payments-namespace)
        WorkflowServiceStubs service = WorkflowServiceStubs.newLocalServiceStubs();
        WorkflowClientOptions clientOptions = WorkflowClientOptions.newBuilder()
                .setNamespace("payments-namespace")
                .build();
        WorkflowClient client = WorkflowClient.newInstance(service, clientOptions);

        // R — Register
        WorkerFactory factory = WorkerFactory.newInstance(client);
        Worker worker = factory.newWorker(Shared.TASK_QUEUE);

        // ┌─────────────────────────────────────────────────────────────┐
        // │ TODO 5 (CHANGE 1): Replace this simple registration with   │
        // │ WorkflowImplementationOptions that map                     │
        // │ "ComplianceNexusService" to "compliance-endpoint"          │
        // │ Include BOTH PaymentProcessingWorkflowImpl.class AND       │
        // │ ReviewCallerWorkflowImpl.class in the same call.           │
        // └─────────────────────────────────────────────────────────────┘
        worker.registerWorkflowImplementationTypes(PaymentProcessingWorkflowImpl.class);

        // A — Activities
        PaymentGateway gateway = new PaymentGateway();
        worker.registerActivitiesImplementations(new PaymentActivityImpl(gateway));

        // ┌─────────────────────────────────────────────────────────────┐
        // │ TODO 5 (CHANGE 2): Delete these two lines after adding     │
        // │ Nexus. Compliance now runs on its own worker.              │
        // └─────────────────────────────────────────────────────────────┘
        ComplianceChecker checker = new ComplianceChecker();
        worker.registerActivitiesImplementations(new ComplianceActivityImpl(checker));

        // L — Launch
        factory.start();

        System.out.println("=========================================================");
        System.out.println("  Payments Worker started on: " + Shared.TASK_QUEUE);
        System.out.println("  Namespace: payments-namespace");
        System.out.println("  Registered: PaymentProcessingWorkflow, PaymentActivity");
        System.out.println("              ComplianceActivity (monolith — will decouple)");
        System.out.println("=========================================================");
    }
}

File: kotlin/decouple-monolith/exercise/src/main/kotlin/payments/temporal/PaymentsWorkerApp.kt
package payments.temporal

import compliance.ComplianceChecker
import compliance.temporal.activity.ComplianceActivityImpl
import io.temporal.client.WorkflowClient
import io.temporal.client.WorkflowClientOptions
import io.temporal.serviceclient.WorkflowServiceStubs
import io.temporal.worker.WorkerFactory
import io.temporal.worker.WorkflowImplementationOptions
import io.temporal.workflow.NexusServiceOptions
import payments.PaymentGateway
import payments.Shared
import payments.temporal.activity.PaymentActivityImpl

/**
 * MONOLITH VERSION. One Worker, one deployment, one blast radius.
 *
 * Right now this single process runs everything:
 *   - PaymentProcessingWorkflow   (Payments team)
 *   - PaymentActivity             (Payments team)
 *   - ComplianceActivity          (Compliance team's code, running in Payments' process)
 *
 * That last line is the problem. A bug in compliance code takes payments down with it.
 *
 * Do TODO 10 after TODOs 8 and 9, which move the Workflow onto a Nexus stub.
 */
fun main() {
    // Client (newInstance): connect, scoped to the Payments team's own Namespace.
    val service = WorkflowServiceStubs.newLocalServiceStubs()
    val client = WorkflowClient.newInstance(
        service,
        WorkflowClientOptions.newBuilder()
            .setNamespace("payments-namespace")
            .build(),
    )

    // Worker (newWorker): poll the Payments Task Queue.
    val factory = WorkerFactory.newInstance(client)
    val worker = factory.newWorker(Shared.TASK_QUEUE)

    // ── TODO 10 ───────────────────────────────────────────────────────────────────
    // Two edits, both on the registration lines below.
    //
    // ONE. Replace the registerWorkflowImplementationTypes line with this, filling in
    // the single blank:
    //
    //     worker.registerWorkflowImplementationTypes(
    //         WorkflowImplementationOptions.newBuilder()
    //             .setNexusServiceOptions(
    //                 mapOf(
    //                     "ComplianceNexusService" to NexusServiceOptions.newBuilder()
    //                         // TODO: point this at the Endpoint "compliance-endpoint"
    //                         .build()
    //                 )
    //             )
    //             .build(),
    //         PaymentProcessingWorkflowImpl::class.java,
    //         ReviewCallerWorkflowImpl::class.java,
    //     )
    //
    // TWO. Delete the ComplianceActivityImpl registration, and the two compliance
    // imports at the top of this file.
    //
    // Stuck on the blank? "Ask AI" on https://docs.temporal.io:
    //   "how do I set the Endpoint on NexusServiceOptions in Java?"
    worker.registerWorkflowImplementationTypes(PaymentProcessingWorkflowImpl::class.java)

    worker.registerActivitiesImplementations(PaymentActivityImpl(PaymentGateway()))

    worker.registerActivitiesImplementations(ComplianceActivityImpl(ComplianceChecker()))

    // Start (start): begin polling.
    factory.start()

    println("=========================================================")
    println("  Payments Worker started on: ${Shared.TASK_QUEUE}")
    println("  Namespace: payments-namespace")
    println("  Registered: PaymentProcessingWorkflow, PaymentActivity")
    println("              ComplianceActivity (monolith, will decouple)")
    println("=========================================================")
}
